src/assets/formats/textures.rs

```diff
@@ -0,0 +1,89 @@
+//! Provides texture formats
+
+use std::mem::replace;
+use std::sync::{Arc, Mutex};
+use std::thread::spawn;
+
+use assets::Format;
+use futures::{Async, Future};
+use image;
+use image::{ImageBuffer, ImageFormat, Rgba};
+pub use image::ImageError;
+
+/// ImageData provided by formats, can be interpreted as a texture.
+#[derive(Clone, Debug)]
+pub struct ImageData {
+    raw: ImageBuffer<Rgba<u8>, Vec<u8>>,
+}
+
+/// A future which will eventually have an image available.
+pub struct ImageFuture {
+    result: Arc<Mutex<Option<Result<ImageData, ImageError>>>>,
+}
+
+impl Future for ImageFuture {
+    type Item = ImageData;
+    type Error = ImageError;
+
+    fn poll(&mut self) -> Result<Async<ImageData>, ImageError> {
+        let mut lock = self.result.lock().unwrap();
+        if lock.is_some() {
+            if lock.as_ref().unwrap().is_ok() {
+                let image_data = replace(&mut *lock, None);
+                Ok(Async::Ready(image_data.unwrap().unwrap()))
+            } else {
+                let image_error = replace(&mut *lock, None);
+                Err(image_error.unwrap().unwrap_err())
+            }
+        } else {
+            Ok(Async::NotReady)
+        }
+    }
+}
+
+fn parse_jpeg(bytes: Vec<u8>) -> ImageFuture {
+    let result = Arc::new(Mutex::new(None));
+    let result_clone = result.clone();
+    spawn(move || {
```

Comment:
Don't use `thread::spawn`. If you need to start long-run operation - use `ThreadPool` instead.